### PR TITLE
ONNX Parser

### PR DESCRIPTION
**Pull request by gramalingam**
Introduce
(a) A text-based syntax for ONNX models, 
(b) A parser,
(c) A pretty-printer, and
(d) Test-cases

The parser will help with a couple of use-cases: it provides a simpler syntax for defining functions; it also makes it easier to specify test-cases that generate models for testing various aspects.

For example, a toy model can be specified as below:
```
<
  ir_version: 7,
  opset_import: [ "" : 10 ]
>
agraph (float[N, 128] X, float[128,10] W, float[10] B) => (float[N] C)
{
    T = MatMul(X, W)
    S = Add(T, B)
    C = Softmax(S)
}
```

### CODE REVIEW

**Comment by linkerzhang** [comment]
May I know the motivation of introducing the text based syntax and associated parser please? Is this for model visualization? Thank you very much!

**Comment by gramalingam** [comment]
@linkerzhang : as mentioned in the above description, there are at least two motivations. One is that this serves as an extension of https://github.com/onnx/onnx/blob/8e99dcff98f32ab30fa31ec1fdbdc5eb0df11f04/onnx/defs/function.h#L29 to help simplify the definition of function bodies. Currently, it is too verbose. Second, it will help write test-cases the generate models fairly easily. For example, the value of this can be seen in how test-cases are written in MLIR/LLVM, where input and output are described in text.

**Comment by gramalingam** [comment]
@linkerzhang : do you have any further comments or feedback? Does this make sense? Thanks!

File: onnx/defs/parser.h
Comment (askhade) [inline onnx/defs/parser.h:47]: Can you use data_type_utils here?

File: onnx/defs/printer.cc
Comment (askhade) [inline onnx/defs/printer.cc:76]: const auto& ?

File: onnx/defs/printer.cc
Comment (askhade) [inline onnx/defs/printer.cc:102]: Case where attrproto contains a graphproto or tensorproto

File: onnx/test/cpp/parser_test.cc
Comment (askhade) [inline onnx/test/cpp/parser_test.cc:59]: curious - will x=0.625f work?

File: onnx/test/cpp/parser_test.cc
Comment (askhade) [inline onnx/test/cpp/parser_test.cc:166]: can you also add checks to verify the type and shape of inputs and outputs

**Comment by askhade** [comment]
overall looks good to me... can you add a document describing the text syntax with some examples for writing test cases, functions etc...

File: onnx/defs/printer.cc
Comment (gramalingam) [reply onnx/defs/printer.cc:76]: I think this should be fine for a primitive type like int/float. There is no efficiency gain using a reference to int/float vs using the actual value.

Comment (gramalingam) [reply onnx/defs/printer.cc:76]: For other cases, like strings (or not-yet-handled tensors/graphs), I agree a const-reference should be used.

File: onnx/test/cpp/parser_test.cc
Comment (gramalingam) [reply onnx/test/cpp/parser_test.cc:59]: Good question. At this point, attributes support only int or float. So, we can use decimal points to distinguish between the two, and I think that is sufficient. (If we ever have multiple precisions for floats in attributes, we could extend this, but I don't see any demand for that now.)

Comment (gramalingam) [reply onnx/test/cpp/parser_test.cc:59]: But for tensors, it could make sense. Depends on the syntax for tensors (whether the type should be specified separately or not). May be worth thinking about.

File: onnx/defs/parser.h
Comment (postrational) [inline onnx/defs/parser.h:372]: Should these methods be part of the `public` interface?

Comment (gramalingam) [reply onnx/defs/parser.h]: I made some of them private methods, and left some of them as public. In general, I think code constructing parts of a model are common. Not all users are going to construct a complete-model in one go. So, I made what seems a reasonable classification for now. Thanks!

File: onnx/defs/parser.h
Comment (tomdol) [inline onnx/defs/parser.h:37]: This macro looks kina hacky. You can achieve the same result by using a variadic template or some other construct that will be type safe.

File: onnx/defs/parser.h
Comment (tomdol) [inline onnx/defs/parser.h:43]: I understand that usage of this macro lets you avoid the if-statements at the call site but this is not a very modern-cpp approach. You could even switch to exceptions instead.

File: onnx/defs/parser.h
Comment (tomdol) [inline onnx/defs/parser.h:160]: In situations like this it is preferred to use `++line` rather than `line++`. The same applies to the for-loop statements increasing a counter/pointer.

**Review by tomdol — COMMENTED** [review]
2 general comments:
- all macros used in this PR could be replaced with a type-safe features from more modern C++
- a lot of code from `parser.h` could be moved to `parser.cpp` (the implementation of ParserBase in particular)

Reviewed per request of @postrational

File: onnx/defs/parser.cc
Comment (mbencer) [inline onnx/defs/parser.cc:148]: Saving uinit64_t as int64_t can cause potential numeric overflow

File: onnx/defs/parser.h
Comment (gramalingam) [reply onnx/defs/parser.h]: Done. But it is a pity if a compiler cannot optimize this in this day and age :)

File: onnx/defs/parser.h
Comment (gramalingam) [reply onnx/defs/parser.h:43]: Yes, but ONNX also supports a build-option to build without exceptions. All things considered, it seems better to avoid exceptions for parser-errors.

File: onnx/defs/parser.cc
Comment (gramalingam) [reply onnx/defs/parser.cc:149]: Replaced.

File: onnx/defs/parser.h
Comment (gramalingam) [reply onnx/defs/parser.h:37]: Are you referring to MakeString? I am reusing what's being used elsewhere in ONNX. The outer macro was just to hide whether exceptions or status is being used to handle errors. The original implementation did use exceptions, but was switched to use Status due to the overall ONNX goals.

Comment (tomdol) [reply onnx/defs/parser.h:37]: No, I meant the whole outer macro. Inside it has this construct with do..while(0), which I've seen many years ago and it was used to adress some specific problem but in a really hacky way. Can this just become a variadic template that builds an error message?

Comment (gramalingam) [reply onnx/defs/parser.h]: Done.

File: onnx/defs/parser.h
Comment (postrational) [inline onnx/defs/parser.h:357]: I'm still not sure why we don't hide all of these `Parse(...Proto&)` methods as `private`

Comment (gramalingam) [reply onnx/defs/parser.h:357]: (a) My experience has been that developers construct the different kinds of objects (such as a TensorShapeProto, or TypeProto) frequently and a convenient way to construct them is useful.
(b) Some of the function-definitions require constructing a graph in which some parts are dependent on context-information (like an input type T); the flexibility exposed by the above interface will make this easier.
(c) A more minor point is that it is convenient to have unit-test-cases that test parsing of the different components. (We could possibly address these using friend-classes, but given the other reasons, the above seems simpler.)
In summary, I think the advantages of the above outweigh the disadvantages (if any). (Of course, we can revisit this later, but unless we expose this functionality, there is no way for users to test it or try it out and give feedback.)

**Review by postrational — APPROVED** [review]

**Comment by postrational** [comment]
We'll need a review from someone in `onnx/sig-archinfra-approvers`

**Comment by gramalingam** [comment]
@askhade : do you have any further feedback on this? Thanks!

**Review by askhade — APPROVED** [review]

**Pull request merged**
